### PR TITLE
Resque Deltas for TS v1/v2/v3 and Flying Sphinx v1

### PR DESCRIPTION
**Pull request by pat**
I've changed a ton here - essentially, it's a rewrite, for a few reasons:
- I really like the idea of locking, but couldn't get it working reliably in tests, so I pulled that out.
- Not sure how the FlagAsDeletedSet was better than a simple FlagAsDeleted job.
- Generally opted for the simplest implementation possible everywhere.
- Cucumber dropped for a few RSpec acceptance tests instead.

All of this is open to discussion - I know this is a massive departure from what you had in place, and it's got me thinking about locking as something that could be a bigger part of Thinking Sphinx indexing generally... although I'm yet to have any inspiration on the best way to implement that.

### CODE REVIEW

**Comment by agibralter** [comment]
Hi Pat,

Thank you so much for the PR! Sorry I haven't had a chance to take a look
yet: I'm super swamped at the moment, but will try to take a look/give
feedback/merge in the next couple of days.

Best,
Aaron

urtak.com / @urtak http://twitter.com/urtak

**Comment by pat** [comment]
I want to confirm it works smoothly in an app locally and on Heroku with Flying Sphinx - so you can hold off on merging for the moment. Given tests are green, hopefully there's no issues, but perhaps a few small tweaks yet to come.

**Comment by pat** [comment]
Just ran it locally and on Heroku, all fine :+1:

**Comment by mattolson** [comment]
@pat Would you consider this production ready? I'd like to use resque delta with TS3 and Flying Sphinx. Does this work?

**Comment by mattolson** [comment]
@pat I have tried downgrading to TS 2.1, but this gem is not compatible (see #23) so I tried 2.0.14, but then saw in the Flying Sphinx docs that 2.1 is required at a minimum. It works for me locally, but I want to use Flying Sphinx on Heroku. So I'm pretty much stuck here...

**Comment by pat** [comment]
@mattolson I gave this fork a run through on a couple of apps on Heroku with Flying Sphinx, so I think you should be good to use it as a git reference in your Gemfile along with TS 2.1 (or if you want, 3.0.3).

If you hit any issues, you're welcome to hit me up here or through http://support.flying-sphinx.com.

**Comment by mattolson** [comment]
@pat Awesome, I'm doing just that.

**Comment by agibralter** [comment]
Really sorry guys, have been totally slammed at work. @pat I haven't had a chance to review the patches but I can add you to the repo/rubygems to release a new version tomorrow. I'll also try to give it a good look through. Sorry again for the delay—should have tackled this already.

​@pat also, any thoughts on the locking?

**Comment by mattolson** [comment]
@pat It is working for me locally and in production. I will let you know if anything crops up, but so far so good! :clap:

File: lib/thinking_sphinx/deltas/resque_delta/delta_job.rb [hunk 1]
```diff
@@ -14,113 +9,6 @@ class ThinkingSphinx::Deltas::ResqueDelta::DeltaJob
...
-    end
-
-    if block_given?
-      result.each do |ary|
-        yield ary
-      end
-    end
-
-    result
+    ThinkingSphinx::Deltas::IndexJob.new(index).perform
```
Comment (agibralter) [inline lib/thinking_sphinx/deltas/resque_delta/delta_job.rb:12]: Is there no need for an around lock to assure that no two `ThinkingSphinx::Deltas::IndexJob.new(index).perform` are happening at the same time on the same index? Or does this make the assumption that you're only running a single resque worker for delta indexing?

Comment (agibralter) [inline lib/thinking_sphinx/deltas/resque_delta/delta_job.rb:12]: And this locking is different from the second layer of locking that used to be in `lib/thinking_sphinx/deltas/resque_delta/core_index.rb`

**Comment by agibralter** [comment]
Hey, @pat,

So I took a closer look and I think my only concerns have to do with locking and the `FlagAsDeletedSet` behavior. 

The around-perform locking does two things:

1

It ensures that only one delta job per index is processed at a time. This is pretty important (at least I think so) to be able have multiple Resque workers working at the same time. I have a bunch of indices, some of which are much bigger/slower and clog up the system. It's nice to be able to process deltas concurrently. Without the lock, I'm not sure what happens when you get, say, two `user_delta` jobs that get fired off at once and both cause the `indexer` to fire at the same time.

and 2

It ensures that the Resque queues do not get huge: 100s-1000s of jobs being created in a short amount time that are essentially redundant (except for the "flag as deleted" part–more on that later) can be needlessly expensive on a Resque setup. The around-perform-lock removes all other instances of the job from the queue before it runs a job. This means 1000 `User` updates in a row could amount to as few as one Resque job. I actually converted another gem of mine https://github.com/agibralter/ar-resque-counter-cache to use resque-loner instead of the around-perform-lock. Resque-loner, rather than allowing jobs to enqueue and then dequeuing them in a lock (probably needlessly inefficient), ensures that no two duplicate jobs are ever in the queue in the first place. I'm pretty sure that method could be used here.

As for "flag as deleted set"... that's another huge performance issue. When 100s-1000s of delta jobs get fired off in a short amount of time, having that translate to the same number of Resque jobs is incredibly expensive. Perhaps Sidekiq would be a better option for us since it is much more efficient at processing jobs, but for now we've had to make due with Resque. The "flag as deleted set" is batch-processing optimization much like the one I've used in https://github.com/agibralter/ar-resque-counter-cache.

Essentially, many of the quirks you eliminated in this update are performance optimizations that I feel pretty strongly that we need. Do you have any ideas about how you'd like to see these kinds of things incorporated? Should the batch processing be defined by an abstract ThinkingSphinx class?

Again, so sorry for the delay in getting to these issues. I'd love to update to newer versions of TS (connection pooling, etc.), but I think the batch processing is still a priority for us.

**Comment by jrust** [comment]
The branch is working well for me.  @agibralter, I'm wondering if it would be better to offload all the locking to [resque-loner](https://github.com/jayniz/resque-loner) rather than trying to reproduce that functionality here.  resque-loner would ensure that a job with the same payload (e.g. index request for the same delta) is only ever added once to the queue which I think would solve both 1 and 2.  The nice thing about resque-loner is that by including the gem as a dependency you aren't affecting any other resque jobs, since you have to explicitly include the `UniqueJob` module into the job class that needs the functionality.

**Comment by jrust** [comment]
BTW, for now I've added this initializer to my app and its working great at keeping the duplicate jobs from piling up on each other:

``` ruby
require 'ts-resque-delta'
ThinkingSphinx::Deltas::ResqueDelta::FlagAsDeletedJob.send :include, Resque::Plugins::UniqueJob
ThinkingSphinx::Deltas::ResqueDelta::DeltaJob.send :include, Resque::Plugins::UniqueJob
```

**Comment by pat** [comment]
Argh, sorry for neglecting this for a while.

On locking generally: Love the idea, but haven't yet figured out what a generic implementation looks like. Threading and maintaining state are the two main things that make this tricky.

On high-load situations generally: I'm not sure how much this gem (or most gems generally?) should handle inefficient situations from the apps that use it.

With regards to avoiding duplicate jobs - this would be great. It's certainly what ts-delayed-delta does, and I've written a Sidekiq middleware as well to do the same thing (though it's not yet part of my ts-sidekiq-delta rewrite). If it all can be done using resque-loner, then fantastic - though from memory I did try to put that in place and headaches ensued (at least within the test suite, or possibly dependency versions).

With regards to flag as deleted - firstly, these jobs should be super fast, but also: if thousands are being fired off in quick succession, I'd argue that the developer should create a custom job that takes more than one document id (or perhaps we provide it, along with an easy way to queue it up), and they also suspend deltas and then fire off a single delta processing job as well.

At the end of the day, if you can deal with these scenarios within ts-resque-delta directly, fantastic - I dropped it to keep things simple, and I wasn't familiar with all the logic and reasoning behind the code you'd written.

**Comment by pat** [comment]
@jrust appreciate the feedback, confirmation that things work, and that resque-loner is working well for you.

**Comment by mattolson** [comment]
@pat I think the resque-loner solution by @jrust should be included in the documentation somewhere. I ran into the same problem -- two workers on Heroku pulling duplicate delta jobs and running at the same time, resulting in process crashes.

**Comment by gustaflindqvist** [comment]
@pat and @agibralter is this going to be merge in to the main branch?

**Comment by agibralter** [comment]
Sorry for the inactivity–I've been super busy lately so I haven't had a chance to get back into thinking-sphinx related stuff...

@pat, do you want to take over ownership of this repo?

**Comment by pat** [comment]
Aaron, you're welcome to add me as a collaborator on this repo, and I'm happy to work on getting a gem release out if you'd like (you'll need to add me to the gem as well: `gem owner ts-resque-delta -a pat@freelancing-gods.com`).

**Comment by ryansch** [comment]
That sounds like a great idea! :+1:

**Comment by agibralter** [comment]
@pat ok, sorry for the delay! Added you as a collaborator to this repo and added you as a gem owner.

**Comment by pat** [comment]
Took far too long, but finally got v2.0.0 released which includes all the changes in this pull request.

**Comment by agibralter** [comment]
Thank you @pat!

**Comment by mattolson** [comment]
Thanks @pat - can't wait to try it out!

**Pull request merged**